### DISCUSSION TITLE
UserWarning: User provided device_type of 'cuda', but CUDA is not available.

### DISCUSSION THREAD

**Discussion by system1system2** [explosion/spaCy#9571] [Help: Installation] [answered]
Hi. I installed spaCy on an x86 Apple laptop **without a CUDA-capable GPU** as indicated in the documentation:

`pip install -U pip setuptools wheel`
`pip install -U spacy`

The command triggered the installation of all necessary dependencies. In this particular case:` typing-extensions`, `numpy`, `murmurhash`, `cymem`, `catalogue`, `wasabi`, `typer`, `srsly`, `smart-open`, `pyparsing`, `pydantic`, `preshed`, `MarkupSafe`, `blis`, `thinc`, `spacy-legacy`, `pathy`, `packaging`, `jinja2`, `spacy`.

It works fine, but as soon as my Python script starts, I receive the following warning:

```
/usr/local/lib/python3.9/site-packages/torch/autocast_mode.py:141: UserWarning: User provided device_type of 'cuda', but CUDA is not available. Disabling
  warnings.warn('User provided device_type of \'cuda\', but CUDA is not available. Disabling')
```

I did nothing to indicate that this is a CUDA device so I'm a bit puzzled by the outcome. How can I disable the attempt to use CUDA? Thank you

## My Environment
- **spaCy version:** 3.1.4
- **Platform:** macOS-12.0.1-x86_64-i386-64bit
- **Python version:** 3.9.6
- **Pipelines:** en_core_web_trf (3.1.0)

**Comment by polm** [comment, marked as answer]
You seem to be using a transformers pipeline, but you don't mention installing `spacy-transformers`. Do you have `spacy-transformers` installed?

What does your code look like? Are you just calling `spacy.load`?

Also note that while it is technically possible to use the Transformers models without a GPU, it is not recommended. You'll have a better time getting started with the CPU models. If you want to switch to Transformers later you can just swap them out, preferably on a machine with a GPU.

**Reply by system1system2** [reply]
I don't have spacy-transformers installed. The relevant part of the code is this:

```
import spacy
nlp = spacy.load('en_core_web_trf')

with open('input.txt', 'r') as f:
    highlight = f.read()
    doc = nlp(input)

for sent in doc.sents:
    print(sent.text)
```

The rest of the script doesn't use spaCy in any way and there's no other part of it that uses any other library. 

If I understood the documentation correctly, CUDA is used only if, for example, I install spaCy with: `pip install -U spacy[cuda113]` which I didn't. 
I did use the (great) widget to generate the installation instructions, specifying to use the CPU. So I don't get why the system is insisting on looking for a CUDA-enabled GPU. 

The kind of work that this Python script has to do is so light and so infrequent that not using a GPU is fine. I just want to be sure that the warning goes away because it wastes time and forces me to do more work to remove it from the output.

**Reply by polm** [reply]
> I don't have spacy-transformers installed. 

Are you sure? Normally installing the trf model will install `spacy-transformers`, and it can't work without it.

 I don't think anyone on the dev team has tested the model on a non-CUDA GPU, so it's not really a use case we've accounted for. I'm not entirely sure what's happening, but I think that since the system detects a GPU it assumes it should try to use CUDA. [This if statement](https://github.com/explosion/spacy-transformers/blob/97d2b13b94f39b8326d4ff5e46050f2a3f944792/spacy_transformers/layers/hf_shim.py#L112) may be at issue. If all you're getting is a warning and performance looks ok, it should be safe to ignore and can be silenced like other Python warnings.

If you really don't want to deal with the warning, you can use `en_core_web_lg`, which is a CPU model. It should be faster and if you're just doing sentence tokenization you shouldn't see any noticeable difference in accuracy.

**Reply by system1system2** [reply]
Clearly, I'm wrong. A `pip list` reveals `spacy-transformers` installed. I must have not noticed. Sorry for not checking before.

The use of `en_core_web_lg` solves the issue (aka no more warnings).

Thanks for helping.

**Reply by vquilon** [reply]
Comparing both models, I executed TRF model with non-CUDA, and the results are very different from the LG model. I want to use the TRF, without that warning, but I don't want to silence the warnings for a obviously reason. I will take a look to the PyTorch DOC.

**Reply by neumannjan** [reply]
I am pretty sure that `spacy-transformers` can be used without the GPU, at least for evaluation (i.e. using already pretrained models, such as `en_core_web_trf`). Pytorch is literally designed for this anyway. I am doing exactly this and it works completely fine, just with the annoying warning that I can't do anything about.

**Reply by polm** [reply]
Transformers will work on CPU, but it will be extremely slow and impractical for most use cases, so we don't recommend it.

**Comment by procrastinando** [comment]
I had the same error using pytorch 1.10 in CPU, I am using 1.8 LTS instead and I dont have the error anymore.

**Reply by miguelwon** [reply]
Hi,
I just had the same warning. I'm using a fresh installation of `spacy` and `spacy-transformers`. In my case, the error is very likely related to the fact I'm loading a model trained in a machine with a GPU and now I just want to infer in a CPU-only PC.

**Reply by polm** [reply]
@miguelwon Did the model work? It's possible that's just related to some changes in PyTorch and can safely be ignored.

**Reply by miguelwon** [reply]
Ok, thanks. Yes, it worked.

**Comment by msksk** [comment]
Hi. I have a similar question.
I tried to use `spacy-transformers` with `'spacy[apple]'` on M1 MAX and got the same warning.
Is this because `thinc-apple-ops` isn't compatible with transformer models? Is it possible to use M1's acceleration for them?

**Reply by adrianeboyd** [reply]
This warning appears to come from newer versions of torch, but I think it's really just a warning and not an error. There shouldn't be any interaction/conflicts between `torch` and `thinc-apple-ops`.

If you search the pytorch issue tracker, you can find discussions about M1 support, but it sounds like it's mostly still under development.

**Reply by nlovell1** [reply]
Hi, I'm not super familiar with the architecture. I have M1 Max and am also receiving this error. Is my performance affected in any way? Are there any additional settings that need to be changed besides using 'spacy[apple]' to increase performance? How do I ensure my GPU is being used? Spacy seems to be running oddly slow.

**Reply by polm** [reply]
I don't think you need to do anything besides using `spacy[apple]`, but note that Mac M1 GPUs are not supported, see [here](https://github.com/explosion/spaCy/issues/10783#issuecomment-1132523032).

**Comment by thunderlm** [comment]
I made a test.
For sentence split, there are some differences between using ‘en_core_web_lg'and 'en_core_web_trf' .It seems that  'en_core_web_trf' is more accuracy.
An example sentence:
_"I BEGAN WRITING THIS BOOK shortly after the end of my presidency—after Michelle and I had boarded Air Force One for the last time and traveled west for a long-deferred break. The mood on the plane was bittersweet. Both of us were drained, physically and emotionally, not only by the labors of the previous eight years but by the unexpected results of an election in which someone diametrically opposed to everything we stood for had been chosen as my successor. Still, having run our leg of the race to completion, we took satisfaction in knowing that we’d done our very best—and that however much I’d fallen short as president, whatever projects I’d hoped but failed to accomplish, the country was in better shape now than it had been when I’d started. For a month, Michelle and I slept late, ate leisurely dinners, went for long walks, swam in the ocean, took stock, replenished our friendship, rediscovered our love, and planned for a less eventful but hopefully no less satisfying second act. And by the time I was ready to get back to work and sat down with a pen and yellow pad (I still like writing things out in longhand, finding that a computer gives even my roughest drafts too smooth a gloss and lends half-baked thoughts the mask of tidiness), I had a clear outline of the book in my head. "_

trf result：

_1
I BEGAN WRITING THIS BOOK shortly after the end of my presidency—after Michelle and I had boarded Air Force One for the last time and traveled west for a long-deferred break.
2
The mood on the plane was bittersweet.
3
Both of us were drained, physically and emotionally, not only by the labors of the previous eight years but by the unexpected results of an election in which someone diametrically opposed to everything we stood for had been chosen as my successor.
4
Still, having run our leg of the race to completion, we took satisfaction in knowing that we’d done our very best—and that however much I’d fallen short as president, whatever projects I’d hoped but failed to accomplish, the country was in better shape now than it had been when I’d started.
5
For a month, Michelle and I slept late, ate leisurely dinners, went for long walks, swam in the ocean, took stock, replenished our friendship, rediscovered our love, and planned for a less eventful but hopefully no less satisfying second act.
6
And by the time I was ready to get back to work and sat down with a pen and yellow pad (I still like writing things out in longhand, finding that a computer gives even my roughest drafts too smooth a gloss and lends half-baked thoughts the mask of tidiness), I had a clear outline of the book in my head.


lg result:
_1
I BEGAN WRITING THIS BOOK shortly after the end of my presidency—after Michelle and I had boarded Air Force One for the last time and traveled west for a long-deferred break.
2
The mood on the plane was bittersweet.
3
Both of us were drained, physically and emotionally, not only by the labors of the previous eight years but by the unexpected results of an election in which someone diametrically opposed to everything we stood for had been chosen as my successor.
4
Still, having run our leg of the race to completion, we took satisfaction in knowing that we’d done our very best—and that however much
5
I’d fallen short as president, whatever projects I’d hoped but failed to accomplish, the country was in better shape now than it had been when I’d started.
6
For a month, Michelle and I slept late, ate leisurely dinners, went for long walks, swam in the ocean, took stock, replenished our friendship, rediscovered our love, and planned for a less eventful but hopefully no less satisfying second act.
7
And by the time I was ready to get back to work and sat down with a pen and yellow pad (I still like writing things out in longhand, finding that a computer gives even my roughest drafts too smooth a_ gloss and lends half-baked thoughts the mask of tidiness), I had a clear outline of the book in my head._

**Reply by polm** [reply]
@thunderlm Did you mean to make a new thread? It's not clear to me what this has to do with anything else here.

**Reply by TheLustriVA** [reply]
The _lg model was offered as a solution to the warning while using the _trf model. @thunderlm's response shows that _lg is not a solution in that it's not as accurate. I'd say that's worth including for others so that they know they can't just swap the two around and expect the same behaviour in their software.

**Reply by polm** [reply]
Ah, I see, you're referring to this comment:

> If you really don't want to deal with the warning, you can use en_core_web_lg, which is a CPU model. It should be faster and if you're just doing sentence tokenization you shouldn't see any noticeable difference in accuracy.

That's a good point. My intent was to communicate that it should be similar in performance in sentence tokenization across a corpus. But the two pipelines certainly won't always produce the same predictions. If you look at [the model page](https://spacy.io/models/en) you can see both TRF and LG have an F1 of 90 for sentences.
